### PR TITLE
[sw/silicon_creator] shutdown_finalize() updates and hardenings

### PR DESCRIPTION
**Pull request by alphan**
This PR:
* Adds a software reset request to `shutdown_hang()` and modifies the inline assembly to include SRAM scrambling key and reset requests in the loop,
* Updates `shutdown_report_error()` to directly write to UART instead of using `log_printf()`, and
* Removes hamming distance histograms from several comments under `sw/device/silicon_creator`.

### CODE REVIEW

**Comment by alphan** [comment]
[disassembly](https://gist.github.com/8fc523833b452be2292bdb625437a67d)

File: sw/device/silicon_creator/lib/shutdown.c [hunk 1]
```diff
@@ -255,6 +257,60 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
...
+/**
+ * Prints a fixed-length message prefix.
+ *
+ * @param prefix Prefix to print.
+ */
+static ALWAYS_INLINE void shutdown_print_prefix(const char *prefix) {
+  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
+    // `kMsgPrefixLen` is laundered so that it is loaded to a register at every
+    // iteration.
+    if (i >= launder32(kMsgPrefixLen)) {
```
Comment (mcy) [inline sw/device/silicon_creator/lib/shutdown.c:287]: Why is this necessary? Can't you just do this comparison in the loop? Is this intended to be a funny hardening?

File: sw/device/silicon_creator/lib/shutdown.c [hunk 2]
```diff
@@ -255,6 +257,60 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
...
+
+static_assert((kMsgPrefixLen + kHexCharsLen) <= 32,
+              "Total message length must be less than TX FIFO size.");
+
+/**
+ * Prints a fixed-length message prefix.
+ *
+ * @param prefix Prefix to print.
+ */
+static ALWAYS_INLINE void shutdown_print_prefix(const char *prefix) {
```
Comment (mcy) [inline sw/device/silicon_creator/lib/shutdown.c:283]: Would it make more sense for this to take a C string and exit if you hit the maximum prefix length? Since you call this and print32 together, maybe it makes sense to merge them and have the `: 0x` be filled in automatically.

If there's a good reason not to do this, it should be documented.

File: sw/device/silicon_creator/lib/shutdown.c [hunk 3]
```diff
@@ -269,12 +325,40 @@ SHUTDOWN_FUNC(NO_MODIFIERS, shutdown_report_error(rom_error_t reason)) {
...
+  const char kLcStatePrefix[kMsgPrefixLen] = "lc_state  : 0x";
+  shutdown_print_prefix(kLcStatePrefix);
+  shutdown_print32(raw_state);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, '\n');
+
+  // Wait until UART TX is idle.
+  is_empty = false;
+  do {
+    is_empty =
+        bitfield_bit32_read(abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET),
```
Comment (mcy) [inline sw/device/silicon_creator/lib/shutdown.c:359]: Why not
```c
while(bitfield_bit32_read(...)) {}
```

File: sw/device/silicon_creator/lib/shutdown.c (hunk 1 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/shutdown.c:287]: Kind of. The loop already has this check but this is meant to be a redundant check against printing more than we want to, loading `kMsgPrefixLen` at each iteration. PTAL at `bc56` below:

```
launder32():
/workspace/worktrees/opentitan/boot-policy/sw/device/lib/base/hardened.h:242
    bc56:	47b9                	li	a5,14
shutdown_print_prefix():
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:287
    if (i >= launder32(kMsgPrefixLen)) {
    bc58:	00f57a63          	bgeu	a0,a5,bc6c <shutdown_finalize+0x1bc>
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:290
    abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix[i]);
    bc5c:	00d507b3          	add	a5,a0,a3
    bc60:	0007c783          	lbu	a5,0(a5)
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:284
  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
    bc64:	0505                	addi	a0,a0,1
abs_mmio_write32():
/workspace/worktrees/opentitan/boot-policy/sw/device/lib/base/abs_mmio.h:91
  *((volatile uint32_t *)addr) = value;
    bc66:	cf5c                	sw	a5,28(a4)
shutdown_print_prefix():
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:284
    bc68:	feb517e3          	bne	a0,a1,bc56 <shutdown_finalize+0x1a6>
```

File: sw/device/silicon_creator/lib/shutdown.c (hunk 3 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/shutdown.c:359]: It actually looked worse than this with the abs_mmio_read32 but I split it so that the register is read in the loop body.

**Review by moidx — APPROVED** [review]

File: sw/device/silicon_creator/lib/shutdown.c [hunk 4]
```diff
@@ -255,6 +257,81 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
...
+    }
+    abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, kHexTable[nibble]);
+  }
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, '\n');
+
+  // Wait until UART TX is idle.
+  uint32_t status_reg;
+  do {
+    status_reg = abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET);
+  } while (!bitfield_bit32_read(status_reg, UART_STATUS_TXIDLE_BIT));
```
[the remark this answers is not in the document]
Comment (alphan) [reply sw/device/silicon_creator/lib/shutdown.c:332]: Checked the UART HWIP spec again: TXIDLE is true when either the FIFO is non-empty or there is a transmission. Thus this loop blocks until all characters are transmitted (may never happen if UART is not transmitting for any reason). Enabling transmission unconditionally was an attempt at addressing this but a limit on the number of iterations would be better. Determining this constant requires some work because CPU clock frequency and baud rate depends on the device and they are link-time constants. If we just define the constant here, we end up with multiplication and division operations so I'll trying adding a constant like `kUartNCOValue` and push an update shortly.

**Comment by alphan** [comment]
* Changed the prefix to use a `uint32_t`
* Added `kUartTxFifoCpuCycles` to `device.h` to limit the number of iterations while waiting for UART transmission.

[disassembly](https://gist.github.com/3230ed1245b17ad09f0c1928147d4641)

File: sw/device/silicon_creator/lib/shutdown.c [hunk 5]
```diff
@@ -269,12 +323,38 @@ SHUTDOWN_FUNC(NO_MODIFIERS, shutdown_report_error(rom_error_t reason)) {
...
+
+  // Encodes a 4 character prefix as a 32-bit value. The resulting prefix is the
+  // concatenation of the given characters and '_'.
+#define MSG_PREFIX(a_, b_, c_) ('_' << 24 | (c_) << 16 | (b_) << 8 | (a_))
 
-  // TODO(lowRISC/opentitan#8236): log_printf is in the .text section.
-  log_printf("boot_fault: 0x%x\n", (unsigned int)redacted_error);
+  // Print the error message and the raw life cycle state as reported by the
+  // hardware.
+  shutdown_print(MSG_PREFIX('B', 'F', 'V'), redacted_error);
```
Comment (moidx) [inline sw/device/silicon_creator/lib/shutdown.c:339]: Nice!. One more nit: It would be nice to have these log codes in a central location. Maybe a header file?

**Review by moidx — APPROVED** [review]

File: sw/device/silicon_creator/lib/shutdown.c (hunk 5 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/shutdown.c]: Done

File: sw/device/silicon_creator/lib/shutdown.c [hunk 6]
```diff
@@ -255,6 +259,57 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
...
+                                         uint32_t val) {
+  // Print the 4 character `prefix`.
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix >> 8);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix >> 16);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix >> 24);
+
+  // Print the hex representation of `val`.
+  const char kHexTable[16] = "0123456789abcdef";
+  // `kHexStrLen` is laundered so that it is loaded to a register at every
```
Comment (mcy) [inline sw/device/silicon_creator/lib/shutdown.c:303]: Is unrolling this loop a problem? That's the only reason I can see laundering being something we care about here, since there's no loop completion check.
